### PR TITLE
Add encoder wrapper to correct high CPS integer overflow

### PR DESCRIPTION
**Pull request by henopied**
The `Encoder` wrapper provides a system to correct velocity integer overflows using a numerical velocity estimate and the raw velocity counts. This wrapper also makes reversing encoders easier without affecting the underlying motor if the ports are shared.

### CODE REVIEW

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import com.acmerobotics.roadrunner.util.NanoClock;
+import com.qualcomm.robotcore.hardware.DcMotorEx;
+
+/**
+ * Wraps a motor instance to provide corrected velocity counts and allow reversing without changing the corresponding
+ * slot's motor direction
+ */
+public class Encoder {
+    public static int CPS_STEP = 0x10000;
```
Comment (rbrott) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java:11]: `private final`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    public static double inverseOverflow(double input, double estimate) {
```
Comment (rbrott) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java:13]: `private`

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            double currentTime = clock.seconds();
+            double dt = currentTime - lastUpdateTime;
+            velocityEstimate = (currentPosition - lastPosition) / dt;
+            lastPosition = currentPosition;
+            lastUpdateTime = currentTime;
+        }
+        return currentPosition;
+    }
+
+    public double getVelocity() {
```
Comment (rbrott) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java:85]: I would qualify the name so it's obvious that this is compensating for overflow and doesn't match the behavior of `DcMotorEx#getVelocity()`. Having a `getVelocity()` that doesn't compensate would be nice.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/StandardTrackingWheelLocalizer.java [hunk 4]
```diff
@@ -43,9 +43,11 @@ public StandardTrackingWheelLocalizer(HardwareMap hardwareMap) {
                 new Pose2d(FORWARD_OFFSET, 0, Math.toRadians(90)) // front
         ));
 
-        leftEncoder = hardwareMap.get(DcMotorEx.class, "leftEncoder");
-        rightEncoder = hardwareMap.get(DcMotorEx.class, "rightEncoder");
-        frontEncoder = hardwareMap.get(DcMotorEx.class, "frontEncoder");
+        leftEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "leftEncoder"));
+        rightEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "rightEncoder"));
+        frontEncoder = new Encoder(hardwareMap.get(DcMotorEx.class, "frontEncoder"));
```
Comment (rbrott) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/StandardTrackingWheelLocalizer.java:48]: I'm leaning toward the overflow compensation change being opt in for now, although I'm fine with the use of `Encoder` for independent direction control. I don't really care whether there's a class-level flag or comment in `getWheelVelocities()`.

**Comment by henopied** [comment]
The numerical screenshot is from the Control Hub, but i suspect that the numerical method may be more noisy on the expansion hub + phone, so I would consider adding a method to average the last few velocities to reduce the noise.
raw
numerical
corrected

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java [hunk 5]
```diff
@@ -82,7 +82,12 @@ public int getCurrentPosition() {
...
 
-    public double getVelocity() {
+    public double getRawVelocity() {
+        int multiplier = direction.getMultiplier();
+        return motor.getVelocity() * multiplier;
+    }
+
+    public double getCorrectedVelocity() {
         int multiplier = direction.getMultiplier();
         double rawVelocity = motor.getVelocity() * multiplier;
```
Comment (rbrott) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Encoder.java]: Nit: this method should just call `getRawVelocity()`.

**Comment by rbrott** [comment]
I'm ready to merge if there are no further changes.

**Comment by henopied** [comment]
I think everything is good here unless you want me to smooth the numerical derivative somehow.

**Comment by rbrott** [comment]
Smoothing can wait. This looks good for now.

**Pull request merged**